### PR TITLE
Renamed lazy foldRight to foldLazy, and made it stack-safe.

### LINKED ISSUE #88: Stack safety of lazy version of `foldRight` for `List`
Cats defines a lazy version of `foldRight` on `Foldable`:
```scala
def foldRight[A, B](fa: List[A], b: Lazy[B])(f: (A, Lazy[B]) => B): Lazy[B] = {
```

Currently the [`List` implementation](https://github.com/non/cats/blob/0e4e235758645c432c5652f2653d0723c21e0773/std/src/main/scala/cats/std/list.scala#L43) is properly lazy, but it is not stack safe.

For example, consider the following derived `Foldable` method:
```scala
def exists[A](fa: F[A])(f: A => Boolean): Boolean =
  foldRight(fa, Lazy.eager(false))((a, b) => f(a) || b.force).force
```

This implementation will properly short-circuit for a small list:
```scala
scala> Foldable[List].exists(List(1, 2, 3)){x => println(x); x > 1}
1
2
res0: Boolean = true
```

However, for a large list it will throw a `StackOverflowError`. If a `List` is long enough that evaluating the predicate for each element is bad (and thus the `Lazy` variant is preferred), then it is likely large enough to throw this error. Therefore something needs to give.

Solutions that come to mind:
1. Use a version of `foldRight` for `List` that uses a combination of `reverse` and `foldLeft`. This would sacrifice laziness.
2. Use trampolining in the `List` implementation of the `Lazy` variant of `foldRight`. The trampolining would add significant overhead, but laziness would be preserved.

The second version seems preferable to me. If somebody is choosing to explicitly call the lazy (as opposed to strict) version, then they should get laziness.

However, I can also see the possibility of the abstraction becoming somewhat leaky. That is, someone defining `def foo[F[_]: Foldable]...` may start to think about which kinds of `F` people are going to be passing into this method so they can decide whether they want potential overhead of laziness or not. Having said that, it seems to me that the abstraction is still pretty leaky (or flat out wrong) if the types are suggesting they will get lazy evaluation but they don't.

What do people think? Beyond just choosing option 1 or option 2, are there other possibilities that come to mind?

### PR DESCRIPTION
**Pull request by non**
This method introduces a new data type, Fold[A]. This type
allows to programmer to decide whether to continue evaluating
the fold (and provide an appropriate continuation function),
or to return immediately.

This fixes #88

### CODE REVIEW

**Comment by non** [comment]
There are a few things I'd like feedback on with this PR:

1. Is the new name (`foldLazy`) good or bad? The overloading of `foldRight` was not great -- in particular it caused problems with implicit syntax. We could alternately name it some variation of `foldRight`.

2. How do you feel about `Fold[_]`? Is the `Pass` hack not worth the trouble? (It will only be useful when implementing methods like `exists`, `forall`, `find`, and so on.) Are the names OK?

3. Do you like the overall method? I feel like this is a pretty elegant way to encode the benefits of `foldr` in Scala, while maintaining stack safety. What do you think?

File: core/src/main/scala/cats/Fold.scala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  /**
+   * iteratRight provides a sample right-fold for Iterable[A] values.
+   */
+  def iterateRight[A, B](as: Iterable[A], b: Lazy[B])(f: A => Fold[B]): Lazy[B] = {
+    def unroll(b: B, fs: List[B => B]): B =
+      fs.foldLeft(b)((b, f) => f(b))
+    def loop(it: Iterator[A], fs: List[B => B]): B =
+      if (it.hasNext) {
+        val a: A = it.next
+        val fb: Fold[B] = f(it.next)
```
Comment (xuwei-k) [inline core/src/main/scala/cats/Fold.scala]: `val fb: Fold[B] = f(a)` ?

Comment (non) [inline core/src/main/scala/cats/Fold.scala]: Oh wow, great catch! What a horrible bug :/

Comment (non) [inline core/src/main/scala/cats/Fold.scala]: I was fooling around with variance annotations and pulled everything out to give manual type annotations -- obviously I failed to clean things up sufficiently.

**Comment by non** [comment]
As @xuwei-k's bug report shows, I need to add some tests to exercise this code.

File: core/src/main/scala/cats/Foldable.scala [hunk 2]
```diff
@@ -20,9 +20,13 @@ import simulacrum._
   def foldRight[A, B](fa: F[A], b: B)(f: (A, B) => B): B
 
   /**
-   * Left for issue #62
+   * Right associative lazy fold on `F` using the folding function 'f'.
+   *
+   * This method evaluates `b` lazily (in some cases it will not be
+   * needed), and returns a lazy value. For more information about how
+   * this method works see the documentation for Fold[_].
```
Comment (julien-truffaut) [inline core/src/main/scala/cats/Foldable.scala]: I think you can add a "@see" for the scaladoc

File: core/src/main/scala/cats/Foldable.scala [hunk 3]
```diff
@@ -20,9 +20,13 @@ import simulacrum._
   def foldRight[A, B](fa: F[A], b: B)(f: (A, B) => B): B
 
   /**
-   * Left for issue #62
+   * Right associative lazy fold on `F` using the folding function 'f'.
+   *
+   * This method evaluates `b` lazily (in some cases it will not be
```
Comment (julien-truffaut) [inline core/src/main/scala/cats/Foldable.scala]: maybe we could add a comment saying why are we using `Fold`, if I understood correctly it is to implement a *stack-safe* lazy foldRight

**Comment by ceedubs** [comment]
@non Good work; I like this. Scalaz has had lots of issues with things like exists and forall not short-circuiting. In fact it looks like the rest (hopefully) were finally cleaned up just 2 days ago: https://github.com/scalaz/scalaz/pull/896 (thanks, @xuwei-k!).

I think using this approach as the default on derived `Foldable` methods could really help prevent that sort of thing.

I'll hold off on a thumbsup until you've had a chance to write the tests you mentioned, but I wanted to say thanks for putting this together and let you know that I think it looks like a good approach.

**Comment by non** [comment]
@ceedubs Thanks for the feedback! I have some basic tests in, so I feel pretty good about this now.

File: core/src/main/scala/cats/Foldable.scala (hunk 3 above)
Comment (non) [inline core/src/main/scala/cats/Foldable.scala]: That's a good point. I'll add some better language there.

**Comment by ceedubs** [comment]
:+1:

File: core/src/main/scala/cats/Fold.scala [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ *  - res2 produces 0
+ *  - res1(0) produces 1
+ *  - res0(1) produces 2
+ *
+ * And so the result is 2.
+ *
+ * This code searches an infinite stream for 77:
+ *
+ *    val found: Lazy[Boolean] =
```
Comment (tixxit) [inline core/src/main/scala/cats/Fold.scala]: A bit pedantic, but if we wrap the docs here in {{{ ... }}} then ScalaDoc will treat it as preformatted text.

Comment (non) [inline core/src/main/scala/cats/Fold.scala]: Oh, that's a good point.

**Comment by tixxit** [comment]
:+1:

**Comment by stew** [comment]
:+1: 

Thanks for all the hard work on this @non, I think this is fantastic

**Pull request merged**